godot-core/src/builtin/rect2i.rs

```diff
@@ -78,21 +82,492 @@ impl Rect2i {
+    ///
+    /// If `size` is an odd number, the returned center value will be rounded towards `position`.
+    #[inline]
+    pub fn get_center(&self) -> Vector2i {
+        self.position + (self.size / 2)
+    }
+
+    /// Returns a copy of the `Rect2i` grown by the specified `amount` on all sides.
+    #[inline]
+    pub fn grow(self, amount: i32) -> Self {
+        let amount_2d = Vector2i::new(amount, amount);
+        Self::from_corners(self.position - amount_2d, self.end() + amount_2d)
+    }
+
+    /// Returns a copy of the `Rect2i` grown by the specified amount on each side individually.
+    #[inline]
+    pub const fn grow_individual(self, left: i32, top: i32, right: i32, bottom: i32) -> Self {
+        Self::from_components(
+            self.position.x - left,
+            self.position.y - top,
+            self.size.x + left + right,
+            self.size.y + top + bottom,
+        )
+    }
+
+    /// Returns a copy of the `Rect2i` grown by the specified `amount` on the specified `RectSide`.
+    #[inline]
+    pub const fn grow_side(self, side: RectSide, amount: i32) -> Self {
+        match side {
+            RectSide::Left => self.grow_individual(amount, 0, 0, 0),
+            RectSide::Top => self.grow_individual(0, amount, 0, 0),
+            RectSide::Right => self.grow_individual(0, 0, amount, 0),
+            RectSide::Bottom => self.grow_individual(0, 0, 0, amount),
+        }
+    }
+
+    /// Returns `true` if the `Rect2i` has area, and `false` if the `Rect2i` is linear, empty, or
+    /// has a negative `size`.
+    #[inline]
+    pub const fn has_area(&self) -> bool {
+        self.size.x > 0 && self.size.y > 0
+    }
+
+    /// Returns `true` if the `Rect2i` contains a point. By convention, the right and bottom edges
+    /// of the `Rect2i` are considered exclusive, so points on these edges are not included.
+    #[inline]
+    pub const fn has_point(&self, point: Vector2i) -> bool {
+        self.assert_nonnegative();
+        // FIXME: provide an unsafe version without the check?
+        // FIXME: or just remove the check altogether?
+
+        let own_end = self.end();
+        point.x >= self.position.x
+            && point.y >= self.position.y
+            && point.x < own_end.x
+            && point.y < own_end.y
+    }
+
+    /// Returns the intersection of this `Rect2i` and `b`.
+    ///
+    /// If the rectangles do not intersect, an empty `Rect2i` is returned.
+    #[inline]
+    pub fn intersection(self, b: Self) -> Self {
+        if !self.intersects(b) {
+            return Self::ZERO;
+        }
+
+        let new_x = cmp::max(b.position.x, self.position.x);
+        let new_y = cmp::max(b.position.y, self.position.y);
+
+        let own_end = self.end();
+        let b_end = b.end();
+        let new_width = cmp::min(b_end.x, own_end.x) - new_x;
+        let new_height = cmp::min(b_end.y, own_end.y) - new_y;
+
+        Self::from_components(new_x, new_y, new_width, new_height)
+    }
```

Comment:
Please return `Option::None` in the error case. While GDScript has to abuse degenerate values for this case, Rust gives us that type safety 🙂

Might make more sense to move the `intersects()` logic here, possibly some double-computations can be avoided.

This method would also be a candidate to use `Vector2i::cwise_max/min()`, as well as `Self::from_corners()` instead of subtractions.

Last, we might want to check for non-negativity of both rectangles (the behavior can be quite unintuitive and is usually unintended).